pyro/__init__.py

```diff
@@ -208,36 +206,64 @@ def iarange(name, size=None, subsample_size=None, subsample=None, use_cuda=None)
         schemes. If specified, then `subsample_size` will be set to
         `len(subsample)`.
     :type subsample: Anything supporting `len()`.
+    :param int dim: An optional dimension to use for this independence index.
+        If specified, ``dim`` should be negative, i.e. should index from the
+        right. If not specified, ``dim`` is set to the rightmost dim that is
+        left of all enclosing ``iarange`` contexts.
     :param bool use_cuda: Optional bool specifying whether to use cuda tensors
         for `subsample` and `log_pdf`. Defaults to `torch.Tensor.is_cuda`.
-    :return: A context manager yielding a single 1-dimensional `torch.Tensor`
-        of indices.
+    :return: A reusabe context manager yielding a single 1-dimensional
+        :class:`torch.Tensor` of indices.
 
     Examples::
 
         # This version simply declares independence:
         >>> with iarange('data'):
-                observe('obs', normal, data, mu, sigma)
+                sample('obs', Normal(mu, sigma), obs=data)
 
         # This version subsamples data in vectorized way:
         >>> with iarange('data', 100, subsample_size=10) as ind:
-                observe('obs', normal, data.index_select(0, ind), mu, sigma)
+                sample('obs', Normal(mu, sigma), obs=data[ind])
 
         # This wraps a user-defined subsampling method for use in pyro:
         >>> ind = my_custom_subsample
         >>> with iarange('data', 100, subsample=ind):
-                observe('obs', normal, data.index_select(0, ind), mu, sigma)
+                sample('obs', Normal(mu, sigma), obs=data[ind])
+
+        # This reuses two different independence contexts.
+        >>> x_axis = iarange('outer', 320, dim=-1)
+        >>> y_axis = iarange('outer', 200, dim=-2)
+        >>> with x_axis:
+                x_noise = sample("x_noise", Normal(mu, sigma).reshape([320]))
+        >>> with y_axis:
+                y_noise = sample("y_noise", Normal(mu, sigma).reshape([200, 1]))
+        >>> with x_axis, y_axis:
+                xy_noise = sample("xy_noise", Normal(mu, sigma).reshape([200, 320]))
 
     See `SVI Part II <http://pyro.ai/examples/svi_part_ii.html>`_ for an
     extended discussion.
     """
-    size, subsample_size, subsample = _subsample(name, size, subsample_size, subsample, use_cuda)
-    if not am_i_wrapped():
-        yield subsample
-    else:
-        with poutine.scale(None, size / subsample_size):
-            with poutine.indep(name, vectorized=True, size=subsample_size):
-                yield subsample
+    def __init__(self, name, size=None, subsample_size=None, subsample=None, dim=None, use_cuda=None):
+        if dim is not None and dim >= 0:
+            raise ValueError('Expected dim < 0 to index from the right, actual {}'.format(dim))
+        self.name = name
+        self.dim = dim
+        self.size, self.subsample_size, self.subsample = _subsample(name, size, subsample_size, subsample, use_cuda)
```

Comment:
Is it always the case that we'll want an `iarange` instance to use the same subsampled indices?  I'm pretty sure it is, just want to double-check.

Reply:
Yes, we want each `iarange` to fix a single subsample for each trace, so that subsampling agrees among tensors. For example, consider adding `noise = x_noise + y_noise + xy_noise` below:
```py
def model(image):
    x_axis = iarange('outer', 320, dim=-1)
    y_axis = iarange('outer', 200, dim=-2)
    with x_axis:
        x_noise = sample("x_noise", Normal(mu, sigma).reshape([320]))
    with y_axis:
        y_noise = sample("y_noise", Normal(mu, sigma).reshape([200, 1]))
    with x_axis, y_axis:
        xy_noise = sample("xy_noise", Normal(mu, sigma).reshape([200, 320]))
        noise = x_noise + y_noise + xy_noise  # <--- these need to refer to the same indices
        pyro.observe("obs", dist.Normal(noise, 0), data=image)
```
These already broadcast correctly, but we need the subsampling to agree so that the `x` agrees between `x_noise[x]` and `xy_noise[:, x]`.